core/trino-main/src/main/java/io/trino/server/NodesResource.java

```diff
@@ -0,0 +1,389 @@
+
+import com.fasterxml.jackson.annotation.JsonCreator;
+import com.fasterxml.jackson.annotation.JsonProperty;
+import com.google.common.base.Joiner;
+import com.google.common.collect.ImmutableSet;
+import com.google.common.util.concurrent.FutureCallback;
+import com.google.common.util.concurrent.Futures;
+import io.airlift.http.client.BodyGenerator;
+import io.airlift.http.client.HttpClient;
+import io.airlift.http.client.HttpClient.HttpResponseFuture;
+import io.airlift.http.client.Request;
+import io.airlift.http.client.StaticBodyGenerator;
+import io.airlift.http.client.StatusResponseHandler;
+import io.airlift.http.client.StatusResponseHandler.StatusResponse;
+import io.airlift.log.Logger;
+import io.trino.metadata.AllNodes;
+import io.trino.metadata.DiscoveryNodeManager;
+import io.trino.metadata.InternalNode;
+import io.trino.metadata.NodeState;
+import io.trino.server.security.ResourceSecurity;
+
+import javax.annotation.Nullable;
+import javax.annotation.PostConstruct;
+import javax.annotation.PreDestroy;
+import javax.inject.Inject;
+import javax.ws.rs.Consumes;
+import javax.ws.rs.GET;
+import javax.ws.rs.PUT;
+import javax.ws.rs.Path;
+import javax.ws.rs.Produces;
+import javax.ws.rs.core.Response;
+
+import java.net.URI;
+import java.net.URISyntaxException;
+import java.nio.charset.Charset;
+import java.util.ArrayList;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Set;
+import java.util.TreeMap;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicInteger;
+
+import static com.google.common.collect.ImmutableSet.toImmutableSet;
+import static com.google.common.collect.Sets.difference;
+import static com.google.common.util.concurrent.MoreExecutors.directExecutor;
+import static io.airlift.concurrent.Threads.threadsNamed;
+import static io.airlift.http.client.HttpUriBuilder.uriBuilderFrom;
+import static io.airlift.http.client.Request.Builder.preparePut;
+import static io.trino.metadata.NodeState.ACTIVE;
+import static io.trino.metadata.NodeState.DECOMMISSIONED;
+import static io.trino.metadata.NodeState.DECOMMISSIONING;
+import static io.trino.metadata.NodeState.INACTIVE;
+import static io.trino.metadata.NodeState.SHUTTING_DOWN;
+import static io.trino.server.security.ResourceSecurity.AccessType.PUBLIC;
+import static java.util.Objects.requireNonNull;
+import static java.util.concurrent.Executors.newSingleThreadScheduledExecutor;
+import static javax.ws.rs.core.HttpHeaders.CONTENT_TYPE;
+import static javax.ws.rs.core.MediaType.APPLICATION_JSON;
+import static javax.ws.rs.core.MediaType.TEXT_PLAIN;
+
+// NodesResource expose coordinator endpoints to facilitate the auto-scaling of cluster.
+// These endpoints include:
+//   1. /v1/nodes --- list of all alive nodes with NodeState and NodeStatus;
+//   2. /v1/nodes/refreshnodes --- refresh with list of nodes to exclude (decommission);
```

Comment:
I am not convinced we need this one. Can we make it simpler and assume requests to decommision/recommision nodes are done by talking to nodes directly?

Reply:
The idea is to leverage and respect coordinator being the authoritative component that manages worker state. Comparing to directly talk to individual workers, the benefits are:
1. request once to the coordinator with set of nodes to be decommissioned. coordinator will immediately stop scheduling new tasks to the node.
2. coordinator already has communication channel with worker nodes, the external component (autoscaler or cluster manager etc) does not need to directly talk to individual end points of the worker, and rely on individual worker to notify coordinator its decommissioning status, and then coordinator no longer schedule task to the worker. The later is overall less efficient and may have split brain issue by design.
3. ask once the coordinator to retrieve list of nodes and their authoritative states

Reply:
I do no feel strongly agains but i would love another opionion on this one. @findepi / @electrum ?
It is not in-line with interface for scheduling node shutdown - we can do that only by drectly talking to the node.

Reply:
The idea is similar to the graceful decommission support in Hadoop https://hadoop.apache.org/docs/stable/hadoop-yarn/hadoop-yarn-site/GracefulDecommission.html where ResourceManager is like the coordinator. It offers an example but not necessarily a justification of the design as YARN does not have Trino's current way of shutdown worker node concept.

Reply:
So maybe we can get off without it. I do not see a huge benefit of proxying those requestes to worker nodes via coordinator. 
@findepi please share your opinion here.

Reply:
The new ability to decommission/recommission a set of nodes through refreshNodes() against the coordinator endpoint is not in conflict with the existing shutdown request directly to the worker node. Users use the current shutdown (like from a pod's pre-stop hook) can still use it as is. That said, let k8s delete a pod before decommission and rely on pre-stop hook to wait for task completion has limitations so we don't do it in our clusters. It is also compatible with the way that workers start and register into coordinator. The whole idea is almost identical to the Hadoop way of graceful decommission (so is not bizarre and abnormal at least).

The whole feature of this PR is to facilitate automatic and more efficient/intelligent auto-scaling of the cluster from an external component (autoscaler in our case). autoscaler periodically refreshes status of all nodes and recent queries, calculate cluster-wide utilization scores, decide the desired number of pods based on recent usage scores, select workers to decommission/recommission as necessary, and delete decommissioned pods as fit etc. The decisions are based on status of the whole cluster including all workers, rather than a local decision made by one individual worker. 

The refreshNodes() way appears natural and efficient to decommission/recommission a set of nodes than individually talk to each worker directly. It is also one of the key features the PR is to provide and our logic relies on. I don't necessarily expect to be able to convince people but maybe helpful to make my opinion clear at least.

Reply:
Thanks for the clarification. I do not think it is bizzare or abnormal FWIW. Just I am not convinced that proxying state-change requests via coordinator buys us much vs directly talking to worker nodes. And it adds a bit of complexity to Trino which needs to be maintained afterward. If you want to build an autoscaler solution you still need to have an external process that talks to Trino API, and makes decisions should we scale up or down and what to kill. It feels to me like it does not change much for that external process whether it talks just to the coordinator or also to worker nodes. The process will know the worker nodes' endpoints anyway as it polls for worker information.

I would vote for simplifying the coordinator logic and just exposing R/O information, which is needed to make scale-up/down decisions. And assume actuation is done by external process directly talking to workers.

Reply:
Status of all nodes is periodically pulled once from the coordinator through the v1/nodes endpoint instead of polling all workers (200 in our clusters) individually.

Reply:
In our case, talking to a worker failed (even ssh failed, which might be a hardware issue). The worker keeps receiving tasks and causing errors, while we can not do anything. In such case, stopping the worker via coordinator would be very useful @losipiuk @dzhi-lyft @findepi